[LLVM] Factor disabled Libcalls into the initializer

jhuber6 opened the pull request:
Summary:
These Libcalls represent which functions are available to the backend.
If a runtime call is not available, the target sets the the name to
`nullptr`. Currently, this logic is spread around the various targets.
This patch pulls all of the locations that disable libcalls into the
intializer. This patch is effectively NFC.

The motivation behind this patch is that currently the LTO handling uses
the list of all runtime calls to determine which functions cannot be
internalized and must be extracted from static libraries. We do not want
this to happen for libcalls that are not emitted by the backend. A
follow-up patch will move out this logic so the LTO pass can know which
rtlib calls are actually used by the backend.

jhuber6 commented:
@Artem-B We probably want to disable most of these for NVPTX as well. Any clue which ones we need to keep around?

shiltian reviewed (commented):
I think this handling makes it more difficult to maintain. A better alternative would be to have one code block for each target, just as how they are spread in multiple files.

jhuber6 commented:
> I think this handling makes it more difficult to maintain. A better alternative would be to have one code block for each target, just as how they are spread in multiple files.

We have the same sprawl in the `TargetLibraryInfo`. I didn't think it was large enough to justify factoring it out, especially when so many of these are common already.

shiltian commented:
But you did factor them to some degree right? It looks like they are categorized in a mixture of multiple ways. It is neither based on the lib call (like one block for one lib call, or one block for a series of lib calls), nor on the triple, because some triples definitely appear multiple times.

jhuber6 commented:
> But you did factor them to some degree right? It looks like they are categorized in a mixture of multiple ways. It is neither based on the lib call (like one block for one lib call, or one block for a series of lib calls), nor on the triple, because some triples definitely appear multiple times.

Yeah, I just noticed which disabled libcalls were common and made a common branch instead of duplicating it.

shiltian commented:
That's my point. I think it is much easier for each target to "mind their own business" instead of trying to figuring out which ones are common among multiple other targets and which are not such that I'd need to have my own branch. Originally they are distributed in their own file but I can see the motivation and need to move them here. To have one block for each target can make it much cleaner. Some duplication here doesn't matter IMHO.

jhuber6 commented:
> That's my point. I think it is much easier for each target to "mind their own business" instead of trying to figuring out which ones are common among multiple other targets and which are not such that I'd need to have my own branch. Originally they are distributed in their own file but I can see the motivation and need to move them here. To have one block for each target can make it much cleaner. Some duplication here doesn't matter IMHO.

I didn't really anticipate this file changing enough, but I can pull them apart if it makes it easier.

arsenm commented:
I assume the next step is to wrap RuntimeLibcalls in a proper API separated from TargetLowering?

jhuber6 commented:
> I assume the next step is to wrap RuntimeLibcalls in a proper API separated from TargetLowering?

Yes, I'll pull it out so it can be used externally.

jhuber6 committed:
[LLVM] Factor disabled Libcalls into the initializer

Summary:
These Libcalls represent which functions are available to the backend.
If a runtime call is not available, the target sets the the name to
nullptr. Currently, this logic is spread around the various targets.
This patch pulls all of the locations that disable libcalls into the
intializer. This patch is effectively NFC.

The motivation behind this patch is that currently the LTO handling uses
the list of all runtime calls to determine which functions cannot be
internalized and must be extracted from static libraries. We do not want
this to happen for libcalls that are not emitted by the backend. A
follow-up patch will move out this logic so the LTO pass can know which
rtlib calls are actually used by the backend.

efriedma-quic commented:
I suspect long-term, we might want to factor this differently... it's sort of annoying for out-of-tree targets to be forced to modify code outside the target itself.  But it's probably simpler to consider the best way to approach that after everything is unified and pulled out of TargetLowering.  So I'm okay with landing this as-is, and revisiting later.

Artem-B commented:
> @Artem-B We probably want to disable most of these for NVPTX as well. Any clue which ones we need to keep around?

I believe we do disable most/all libcalls, but for the life of me I can't find where we're doing that. I vaguely recall it was in some odd place.

jhuber6 commented:
> > @Artem-B We probably want to disable most of these for NVPTX as well. Any clue which ones we need to keep around?
> 
> I believe we do disable most/all libcalls, but for the life of me I can't find where we're doing that. I vaguely recall it was in some odd place.

So, you're likely confusing the "LLVM-IR" libcalls and the "RTLib" libcalls. I know there's logic in `TargetLibraryInfo` where we disable all except things like `printf`. (Funnily enough the `libc` support should allow us to add these if we want. This is for runtime calls used by the backend.

Artem-B commented:
> you're likely confusing the "LLVM-IR" libcalls and the "RTLib" libcalls. I know there's logic in TargetLibraryInfo where we disable all except things like printf.

Yup. Thank you for clarifying it. In that case, all RTlib calls should be disabled for NVPTX, as there's nothing on the GPU side to fulfill them at the moment, at least for the standard CUDA compilation.

jhuber6 commented:
> Yup. Thank you for clarifying it. In that case, all RTlib calls should be disabled for NVPTX, as there's nothing on the GPU side to fulfill them at the moment, at least for the standard CUDA compilation.

You weren't the only one confused at the duplication here. I can make a patch to turn them off since I'll want it for NVPTX compilation as well pending https://github.com/llvm/llvm-project/pull/96561.
